polars/polars-time/src/chunkedarray/utf8/mod.rs

```diff
@@ -275,29 +277,51 @@ pub trait Utf8Methods: AsUtf8 {
                         if s.is_empty() {
                             return None;
                         }
-                        match NaiveDateTime::parse_from_str(s, fmt).map(func) {
-                            Ok(nd) => return Some(nd),
-                            Err(e) => {
-                                let e: ParseErrorByteCopy = e.into();
-                                match e.0 {
-                                    ParseErrorKind::TooLong => {
-                                        s = &s[..s.len() - 1];
-                                    }
-                                    _ => {
-                                        s = &s[i..];
+                        match tz_aware {
+                            true => {
+                                match DateTime::parse_from_str(s, fmt)
+                                    .map(|dt| func(dt.naive_utc()))
+                                {
+                                    Ok(nd) => return Some(nd),
+                                    Err(e) => {
+                                        let e: ParseErrorByteCopy = e.into();
+                                        match e.0 {
```

Comment:
What are we trying to do here? We return the string slice?

Reply:
if the string is too long (respectively, too short) for the format string, then the end (respectively, the beginning) is chopped off

Reply:
Right, we loop. Nevermind, I didn't expand enough :)

Reply:
TBH this functionality feels like a performance footgun, as it involves repeatedly calling `parse_from_str` for each element

```python
In [64]: dates = 'foo ' + pl.date_range(dt.datetime(2000, 1, 1), dt.datetime(9000, 1, 1), eager=True).dt.strftime('%Y-%m-%d %H:%M') + ' bar'

In [65]: dates.shape
Out[65]: (2556699,)

In [66]: %timeit dates.str.to_datetime(format="%Y-%m-%d %H:%M", exact=False)

3.99 s ± 15.3 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [67]: %timeit dates.str.to_datetime(format="foo %Y-%m-%d %H:%M bar")
981 ms ± 20.5 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [68]: pl.__version__
Out[68]: '0.18.0'
```

In this case when the user knows the prefix/suffix, then on 2.5 million elements the performance is >4x.

It's also kind of buggy because only `NaiveDatetime` (not `NaiveDate`) is tried:
```python
n [69]: pl.Series(['foo2000-01-01bar']).str.to_datetime('%Y-%m-%d', exact=False)
thread '<unnamed>' panicked at 'byte index 7 is out of bounds of `1-01`', /home/runner/work/polars/polars/polars/polars-time/src/chunkedarray/utf8/mod.rs:287:46
---------------------------------------------------------------------------
PanicException: byte index 7 is out of bounds of `1-01`
```
Sorting that out would introduce even more performance penalty

Wondering if this should just be deprecated, and let users clean up their date strings before passing them to `.str.to_datetime`?

Reply:
Let's at least start with a big fat warning in the docstring.

Our algorithm is also very naive, maybe we can come up with better things than just retrying on subslices.